src/Bicep.Core/Emit/ExternalInputFunctionReferenceVisitor.cs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Immutable;
+using System.Reflection.Metadata;
+using Bicep.Core.Semantics;
+using Bicep.Core.Semantics.Namespaces;
+using Bicep.Core.Syntax;
+using Bicep.Core.Syntax.Visitors;
+using Bicep.Core.TypeSystem.Types;
+
+namespace Bicep.Core.Emit;
+
+public sealed class ExternalInputFunctionReferenceVisitor : AstVisitor
+{
+    private readonly SemanticModel semanticModel;
+    private ParameterAssignmentSyntax? targetParameterAssignment;
+    private VariableDeclarationSyntax? targetVariableDeclaration;
+    private readonly ImmutableHashSet<ParameterAssignmentSyntax>.Builder parametersContainingExternalInput;
+    private readonly ImmutableHashSet<VariableDeclarationSyntax>.Builder variablesContainingExternalInput;
+    private readonly ImmutableDictionary<FunctionCallSyntax, int>.Builder externalInputReferences;
+    private ExternalInputFunctionReferenceVisitor(SemanticModel semanticModel)
+    {
+        this.semanticModel = semanticModel;
+        this.externalInputReferences = ImmutableDictionary.CreateBuilder<FunctionCallSyntax, int>();
+        this.parametersContainingExternalInput = ImmutableHashSet.CreateBuilder<ParameterAssignmentSyntax>();
+        this.variablesContainingExternalInput = ImmutableHashSet.CreateBuilder<VariableDeclarationSyntax>();
+    }
+
+    public override void VisitVariableAccessSyntax(VariableAccessSyntax syntax)
+    {
+        var symbol = semanticModel.GetSymbolInfo(syntax);
+
+        if (symbol is DeclaredSymbol declaredSymbol && declaredSymbol.Type is ErrorType)
+        {
+            return;
+        }
+
+        switch (symbol)
+        {
+            case ParameterAssignmentSymbol parameterAssignmentSymbol:
+                Visit(parameterAssignmentSymbol.DeclaringSyntax);
+                break;
+
+            case VariableSymbol variableSymbol:
+                Visit(variableSymbol.DeclaringSyntax);
+                break;
+        }
+
+        base.VisitVariableAccessSyntax(syntax);
+    }
+
+
+    public override void VisitFunctionCallSyntax(FunctionCallSyntax syntax)
+    {
+        if (string.Equals(syntax.Name.IdentifierName, LanguageConstants.ExternalInputBicepFunctionName, LanguageConstants.IdentifierComparison))
```

Comment:
To be more precise, I would use:

```csharp
SemanticModelHelper.TryGetNamedFunction(semanticModel, SystemNamespaceType.BuiltInName, LanguageConstants.ExternalInputBicepFunctionName, syntax)
```

Note that you'll probably also want to implement `VisitInstanceFunctionCallSyntax` and share this logic, as users could also write:
```bicep
var val = sys.externalInput(...)
```

Reply:
Thanks for callout to add `VisitInstanceFunctionCallSyntax`! Updated if check per recommendation.